### PR TITLE
feat: Package and configure Endatix NuGet packages

### LINKED ISSUE #53: Package and configure Endatix NuGet packages
# Description

The following NuGet packages will be added:

### Endatix.Core
core libraries that represent the domain and the framework - it shall include the following DLLs:
* Endatix.Core
* Endatix.Framework

### Endatix.Api
adding the Endatix API to the project. It shall depend on the following package:
* Endatix.Core

and include the following Dlls:
* Endatix.Api
* Endatix.Extensions.Hosting
* Endatix.Infrastructure

### Endatix.SqlServer
adding support for SqlService for data persistence. It shall include the following Dlls:

* Endatix.SqlServer (Rename Endatix.SqlServer to Endatix.Persistence.SqlServer)
* Endatix.Infrastructure

====
# UPDATE
=====
> [!NOTE]
> The

Following Nuget packages were added to follow the recommendations of "single assembly per package" - highly recommended, but not required. We need to review and change if we agree - https://learn.microsoft.com/en-us/nuget/create-packages/creating-a-package-dotnet-cli. Here is the recommendation:

> In general, set the package version to match the project or assembly version, although this isn't strictly required. Matching the version is simple when you limit a package to a single assembly. NuGet itself deals with package versions when resolving dependencies, not assembly versions.

As a result, this is what we got. More info in the related PR

| Package    | Description |
| -------- | ------- |
| Endatix.Api.Host   | Install this package to host the Endatix API - the key presentation layer of the Endatix Platform. Includes Public and Management APIs   |
| Endatix.Api | Key presentation layer of the Endatix Platform. Includes Public and Management APIs    |
| Endatix.Core    | The core of the Endatix Platform. Contains the domain model of the application and common abstractions.    |
|     Endatix.Extensions.Hosting    |    Provides useful extension methods and utilities for hosting the Endatix Platform.     |
|      Endatix.Framework   |    Provides common plugin and extensibility points used to extend or customize the Endatix Platform.     |
|    Endatix.Infrastructure     |    Provides implementations of application's dependencies and external resources.     |
|     Endatix.SqlServer    |   Adds support for using MS SQL Server database persistence with the Endatix Platform.      |
|    Endatix.WebHost     |   NOT A PACKAGE |

### PR DESCRIPTION
**Pull request by PeterMarinov**
# Package and configure Endatix NuGet packages

## Description
* Adding `Directory.Build.props` for common MS Build and Nuget package properties that will be inherited from all projects from the solution. More info [here](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022#directorybuildprops-and-directorybuildtargets)
> You can add a new property to every project by defining it in a single file called Directory.Build.props in the root folder that contains your source.

* Adding `Endatix.Api.Host` project to allow for installing everything needed via single package, e.g.  `dotnet add package Endatix.Api.Host  --prerelease`

* Manual version of 0.1.0-alpha was added. Check the `<PackageVersion>0.1.0-alpha</PackageVersion>` in the `Directory.Build.props` for controlling it

Following Nuget packages were added 1 assembly per package. Read more. - https://learn.microsoft.com/en-us/nuget/create-packages/select-assemblies-referenced-by-projects#recommended-one-assembly-per-package. My personal preference was to have few number of packages, but we can review later. 

Here is the list of current packages:

| Package    | Description |
| -------- | ------- |
| Endatix.Api.Host   | Install this package to host the Endatix API - the key presentation layer of the Endatix Platform. Includes Public and Management APIs   |
| Endatix.Api | Key presentation layer of the Endatix Platform. Includes Public and Management APIs    |
| Endatix.Core    | The core of the Endatix Platform. Contains the domain model of the application and common abstractions.    |
|     Endatix.Extensions.Hosting    |    Provides useful extension methods and utilities for hosting the Endatix Platform.     |
|      Endatix.Framework   |    Provides common plugin and extensibility points used to extend or customize the Endatix Platform.     |
|    Endatix.Infrastructure     |    Provides implementations of application's dependencies and external resources.     |
|     Endatix.SqlServer    |   Adds support for using MS SQL Server database persistence with the Endatix Platform.      |
|    Endatix.WebHost     |   NOT A PACKAGE |

## How to Test
Below is a "draft" of the current setup experience, but first you will need to create a local source. Here are the steps to setup a local source - https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-locals

1. Add new NuGet Source - `dotnet nuget add source /Users/kungfunc/Code/local-nuget-src -n LocalNuget`
2. Validate that new source is added to local NuGet feed - `dotnet nuget list source`
3. CD into the root of the solution and publish the NuGets to a folder named "nuget" - `dotnet pack  --output nuget`
4. Publish packages to the local feed created in step 1 - `dotnet nuget push "nuget/*.nupkg" -s /Users/kungfunc/Code/local-nuget-src`
 
That's it. From there you can test as if the packages are published to NuGet.org. Here are the setup instructions you can test with

# Setup the Site using NuGet packages

### 1. Create a new .NET project
run the command:
`dotnet new web -n endatix-platform`

and then cd into endatix-platform

### 2. Open the project in your IDE

If you use VS Code, you can enter `code .`

### 3. Install the Endatix.Api
run `dotnet add package Endatix.Api.Host  --prerelease` or use NuGet package manager

### 4. Edit Program.cs

Open `Program.cs` file. Add the following code after the builder initialization:

```csharp
builder.CreateEndatix()
    .AddDefaultSetup()
    .AddApiEndpoints();
```

Find where where the app is defined (`var app = builder.Build();`) and below it add the following code, which will register the Endatix middleware:

```csharp
app.UseEndatixMiddleware()
            .UseEndatixApi();
```

At the end your Program.cs should look something like this.

```csharp

using Endatix.Setup;

var builder = WebApplication.CreateBuilder(args);

builder.CreateEndatix()
    .AddDefaultSetup()
    .AddApiEndpoints();

var app = builder.Build();

app.UseEndatixMiddleware()
            .UseEndatixApi();

app.Run();
```

> **NOTE**
> If you receive errors for using prelease APIs, please add this to the csproj file of your project - ` <EnablePreviewFeatures>true</EnablePreviewFeatures>`

### 5. Configure the AppSettings

Endatix uses Serilog with settings from the config. Copy the config and configure your connection string

```json
{
  {
  "ConnectionStrings": {
    "DefaultConnection": "{YOUR_CONNECTION_STRING_HERE}"
  },
  "Serilog": {
    "Using": ["Serilog.Sinks.Console", "Serilog.Sinks.File"],
    "MinimumLevel": {
      "Default": "Debug",
      "Override": {
        "Microsoft": "Information",
        "System": "Information"
      }
    },
    "WriteTo": [
      {
        "Name": "Console",
        "Args": {
          "applyThemeToRedirectedOutput": true,
          "theme": "Serilog.Sinks.SystemConsole.Themes.AnsiConsoleTheme::Sixteen, Serilog.Sinks.Console",
          "outputTemplate": "[{Timestamp: HH:mm:ss.fff} Level:{Level:u3}] {Message:lj}{NewLine}{Exception}"
        }
      },
      {
        "Name": "File",
        "Args": {
          "path": "/logs/log-.txt",
          "rollingInterval": "Day",
          "rollOnFileSizeLimit": true,
          "formatter": "Serilog.Formatting.Json.JsonFormatter"
        }
      }
    ],
    "Enrich": [
      "FromLogContext",
      "WithMachineName",
      "WithProcessId",
      "WithThreadId"
    ],
    "Properties": {
      "Application": "Endatix Local Development",
      "Environment": "Local Development"
    }
  },
  "Security": {
    "JwtSigningKey": "{YOUR_JWT_SIGNING_KEY_HERE}",
    "JwtExpiryInMinutes": 1440,
    "DevUsers": [
      {
        "Email": "{USER_EMAIL}",
        "Password": "{USER_PASSWORD}",
        "Roles": ["Admin", "Manager"]
      }
    ]
  }
}
```

## Related Issues
Closes #53 

## Type of Change
Please delete options that are not relevant.
- [x ] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation - for this check #56 
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by PeterMarinov** [comment]
@dspetrov  - as we have chatted, this is far from perfect, but good for v1 of the process to improve developer experience. I have created this task for a spike to further improve the setup experience - https://github.com/endatix/endatix/issues/74

File: src/Endatix.Api/Setup/EndatixAppExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        endatixApp.Services
+                .AddCors()
+                .AddFastEndpoints()
+                .SwaggerDocument(o =>
+                    {
+                        o.ShortSchemaNames = true;
+                        o.DocumentSettings = s =>
+                        {
+                            s.Version = "v0";
+                            s.DocumentName = "Internal MVP (Alpha) Release";
```
Comment (dspetrov) [inline src/Endatix.Api/Setup/EndatixAppExtensions.cs:29]: A small thing we've commented in a previous PR but somehow it reached here - the internal MVP ;)

File: src/Endatix.Extensions.Hosting/EndatixWebAppBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    /// </summary>
+    /// <param name="builder">The <see cref="WebApplicationBuilder"/> used to configure the application.</param>
+    /// <param name="logger">Optional. The <see cref="Logging.ILogger"/> instance for logging. If null, a default Serilog logger is used.</param>
+    /// <returns>An instance of <see cref="IEndatixApp"/> representing the configured application.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if the <paramref name="builder"/> is null.</exception>
+    public static IEndatixApp CreateEndatix(this WebApplicationBuilder builder, Logging.ILogger? logger = null)
+    {
+        Guard.Against.Null(builder);
+
+        if (logger == null)
```
Comment (dspetrov) [inline src/Endatix.Extensions.Hosting/EndatixWebAppBuilderExtensions.cs:31]: Similar to comments in a previous PR, I think this is a good candidate for method extraction.

**Review by dspetrov — COMMENTED** [review]
I've tested the packages and run the app - it works and is cool, good job!
I've added a couple of comments.

File: src/Endatix.Api/Setup/EndatixAppExtensions.cs (hunk 1 above)
Comment (PeterMarinov) [reply src/Endatix.Api/Setup/EndatixAppExtensions.cs]: @dspetrov  - this was outdated. I have merged latest from `main` and you can see this file has not been changed. Sorry for that.

File: src/Endatix.Extensions.Hosting/EndatixWebAppBuilderExtensions.cs (hunk 2 above)
Comment (PeterMarinov) [reply src/Endatix.Extensions.Hosting/EndatixWebAppBuilderExtensions.cs]: @dspetrov  - this was outdated. I have merged latest from `main` and you can see this file has not been changed. Sorry for that.

**Review by dspetrov — APPROVED** [review]

**Review by dspetrov — APPROVED** [review]
LGTM

**Pull request merged**